File: ts-lambda-worker/src/certs.ts
import { SecretsManagerClient, GetSecretValueCommand } from '@aws-sdk/client-secrets-manager';
import type { TLSConfig } from '@temporalio/worker';

const client = new SecretsManagerClient({});
let cached: TLSConfig | undefined;

async function getSecret(arn: string): Promise<Buffer> {
  const { SecretString } = await client.send(new GetSecretValueCommand({ SecretId: arn }));
  if (!SecretString) throw new Error(`Secret ${arn} has no string value`);
  return Buffer.from(SecretString);
}

/**
 * Fetches mTLS client cert and key from Secrets Manager.
 * Returns undefined if TEMPORAL_TLS_CERT_ARN / TEMPORAL_TLS_KEY_ARN are not set.
 */
export async function getTLSCerts(): Promise<TLSConfig | undefined> {
  const certArn = process.env['TEMPORAL_TLS_CERT_ARN'];
  const keyArn = process.env['TEMPORAL_TLS_KEY_ARN'];
  if (!certArn || !keyArn) return undefined;
  if (cached) return cached;

  const [crt, key] = await Promise.all([getSecret(certArn), getSecret(keyArn)]);
  cached = { clientCertPair: { crt, key } };
  return cached;
}

File: py-lambda-worker/app/certs.py
import os

import boto3
from temporalio.service import TLSConfig

_cached: TLSConfig | None = None
_fetched: bool = False


def get_tls_certs() -> TLSConfig | None:
    """Fetches mTLS client cert and key from Secrets Manager.

    Returns None if TEMPORAL_TLS_CERT_ARN / TEMPORAL_TLS_KEY_ARN are not set.
    Called synchronously at cold start from the run_worker configure callback.
    """
    global _cached, _fetched
    if _fetched:
        return _cached

    cert_arn = os.environ.get("TEMPORAL_TLS_CERT_ARN")
    key_arn = os.environ.get("TEMPORAL_TLS_KEY_ARN")
    if not cert_arn or not key_arn:
        _fetched = True
        return None

    sm = boto3.client("secretsmanager")
    cert = sm.get_secret_value(SecretId=cert_arn)["SecretString"].encode()
    key = sm.get_secret_value(SecretId=key_arn)["SecretString"].encode()
    _cached = TLSConfig(client_cert=cert, client_private_key=key)
    _fetched = True
    return _cached
